### PR TITLE
[red-knot] use reachable definitions in infer_expression_type

### PR DESCRIPTION
**Pull request by carljm**
## Summary

Switch name resolution in `infer_expression_type` from resolving the public type of a symbol, to resolving the reachable definitions of that symbol from the reference point, using the flow graph.

This surfaced a bug in the flow graph implementation and a bug in symbol table building, both of which are also fixed here.

The bug in flow graph implementation was that when we pushed and popped scopes, we didn't maintain a stack of "current flow nodes" in all stacked scopes, to be restored when we returned to that scope. Now we do.

The bug in symbol table building that we didn't visit the parts of functions and class definitions in the correct scopes. E.g. decorators should be visited in the outer scope, arguments should be visited inside the type-params scope (if any) but not inside the function body scope, and only the body itself should actually be visited inside the body scope. Fixing this requires that we no longer use `walk_stmt` here, instead we have to visit each individual component.

## Test Plan

Added test.

### CODE REVIEW

File: crates/red_knot/src/symbols.rs [hunk 1]
```diff
@@ -244,9 +244,8 @@ impl SymbolTable {
         let root_scope_id = SymbolTable::root_scope_id();
         let mut builder = SymbolTableBuilder {
             table: SymbolTable::new(),
-            scopes: vec![root_scope_id],
+            scopes: vec![(root_scope_id, FlowGraph::start())],
```
Comment (MichaReiser) [inline crates/red_knot/src/symbols.rs:247]: Nit: I think I would prefer a named struct here to awoid the awkward `.1` and `.0` which I always find hard to read.

File: crates/red_knot/src/symbols.rs [hunk 2]
```diff
@@ -557,10 +557,13 @@ impl<'a> Iterator for ReachableDefinitionsIterator<'a> {
             match &self.table.flow_graph.flow_nodes_by_id[self.flow_node_id] {
                 FlowNode::Start => return None,
                 FlowNode::Definition(def_node) => {
-                    self.flow_node_id = def_node.predecessor;
                     if def_node.symbol_id == self.symbol_id {
+                        // we found a definition; previous definitions along this path are not
+                        // reachable
+                        self.flow_node_id = FlowGraph::start();
                         return Some(def_node.definition.clone());
```
Comment (MichaReiser) [inline crates/red_knot/src/symbols.rs:564]: Nit: Maybe add an empty line in between to make it clear that the comment only applies to the assignment
```suggestion
                        self.flow_node_id = FlowGraph::start();
                        
                        return Some(def_node.definition.clone());
```

File: crates/red_knot/src/symbols.rs [hunk 3]
```diff
@@ -605,13 +608,21 @@ impl FlowGraph {
...
     /// the definition whose target(s) we are currently walking
     current_definition: Option<Definition>,
-    current_flow_node: FlowNodeId,
 }
 
 impl SymbolTableBuilder {
+    fn set_current_flow_node(&mut self, new_flow_node_id: FlowNodeId) {
+        let (scope_id, _) = self.scopes.pop().expect("scope stack is never empty");
+        self.scopes.push((scope_id, new_flow_node_id));
+    }
```
Comment (MichaReiser) [inline crates/red_knot/src/symbols.rs:620]: The compiler might be clever enough to work this out but maybe
```suggestion
        let (_, flow_node_id) = self.scopes.last_mut().expect("scope stack is never empty");
        *flow_node_id = new_flow_node_id;
    }
```

**Review by MichaReiser — APPROVED** [review]

File: crates/red_knot/src/symbols.rs (hunk 2 above)
Comment (carljm) [reply crates/red_knot/src/symbols.rs:564]: Point taken, but this code changes in next PR anyway; not going to change it here which will just cause conflicts.

**Pull request merged**